### PR TITLE
WIP: New implementations and tests for Zerokit v0.1

### PR DESCRIPTION
**Pull request by s1fr0**
Targets https://github.com/vacp2p/zerokit/issues/31

### CODE REVIEW

**Review by richard-ramos — APPROVED** [review]
:100: 

Looking at the code, a random idea i had that wanted to ask you about is:
What do you think of extracting in a patch file or separate branch these changes?:
- in poseidon_tree.rs:
```rs
//pub type PoseidonTree = FullMerkleTree<PoseidonHash>;
//pub type MerkleProof = FullMerkleProof<PoseidonHash>;
```
- the functions defined in `rln/src/utils.rs`. 

That way the source code would not contain commented code, and someone wanting to use either the FullMerkleTree, or the conversion utils could either cherry pick the commit from the branch, or apply the git patch  (assuming this operation is documented somewhere)

**Comment by richard-ramos** [comment]
Sounds good! I did not know about rust features!

**Review by oskarth — CHANGES_REQUESTED** [review]
This is not a release PR, it is doing _a lot_ of things at once. These should be separate PRs. The release PR should be minimal and only about adding release notes etc.

See e.g. how this is done in nwaku: https://github.com/status-im/nwaku/pull/1075/files

I'm also not convinced that we should decrease the capability this library has as part of the first release.

File: multiplier/Cargo.toml [hunk 1]
```diff
@@ -1,35 +0,0 @@
-[package]
```
Comment (oskarth) [inline multiplier/Cargo.toml:1]: This is a useful example for onboarding and understanding roughly how things fit together. I don't think it makes sense to remove this package.

We could move it to an example directory.

File: poseidon-tornado/README.md [hunk 2]
```diff
@@ -1,15 +0,0 @@
-# Poseidon-Tornado wrapper
```
Comment (oskarth) [inline poseidon-tornado/README.md:1]: We want this soon, but I agree it is in a so-so state and better to base based on RLN from scratch at this point.

Comment (oskarth) [reply poseidon-tornado/README.md:1]: Removed in https://github.com/vacp2p/zerokit/commit/1d8a5f59b60eea88cdd69d68df760ebb3ae73e54

cc @staheri14 if you want to see what this code is doing

File: rln/Cargo.toml [hunk 3]
```diff
@@ -3,76 +3,39 @@ name = "rln"
 version = "0.1.0"
 edition = "2021"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (oskarth) [inline rln/Cargo.toml:6]: Do this in a separate PR

File: rln/src/circuit.rs [hunk 4]
```diff
@@ -1,22 +1,19 @@
 // This crate provides interfaces for the zero-knowledge circuit and keys
 
-use ark_bn254::{Bn254, Fq, Fq2, Fr, G1Affine, G1Projective, G2Affine, G2Projective};
-use ark_circom::{read_zkey, CircomBuilder, CircomConfig, WitnessCalculator};
-use ark_ff::BigInteger256;
+use ark_bn254::{
+    Bn254, Fq as ArkFq, Fq2 as ArkFq2, Fr as ArkFr, G1Affine as ArkG1Affine,
```
Comment (oskarth) [inline rln/src/circuit.rs:4]: This a change in functionality or so, belongs in a separate PR targeting an issue. Not in a release PR...

File: rln/src/merkle_tree.rs [hunk 5]
```diff
@@ -35,239 +32,6 @@ pub trait Hasher {
     fn hash(input: &[Self::Fr]) -> Self::Fr;
 }
 
-////////////////////////////////////////////////////////////
```
Comment (oskarth) [inline rln/src/merkle_tree.rs:38]: This shouldn't be removed, it is a regression

File: rln/src/poseidon_hash.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
+// This crate implements the Poseidon hash algorithm https://eprint.iacr.org/2019/458.pdf
+
+// The implementation is taken from https://github.com/arnaucube/poseidon-rs/blob/233027d6075a637c29ad84a8a44f5653b81f0410/src/lib.rs
+// and slightly adapted to work over arkworks field data type
+
```
Comment (oskarth) [inline rln/src/poseidon_hash.rs:5]: This seems like another feature ??

File: semaphore/Cargo.toml [hunk 7]
```diff
@@ -1,49 +0,0 @@
-[package]
```
Comment (oskarth) [inline semaphore/Cargo.toml:1]: I think this is still useful, we aren't unlikely to want to use this in the feature. There are no problems to keep it in the code base as WIP, and when it gets prioritized to update it.

**Comment by oskarth** [comment]
Yeah I think better to close this PR and start from scratch with feature PRs and refactor PRs and then a minimal release note PR. This is just a monster right now with many separate intentions.

File: rln/Cargo.toml [hunk 8]
```diff
@@ -3,76 +3,39 @@ name = "rln"
 version = "0.1.0"
```
Comment (oskarth) [inline rln/Cargo.toml:3]: @richard-ramos @staheri14  of the changes in the `rln` crate, which of these are must haves from your POV? Both for normal integration (e.g. bugs found in nwaku and go-waku) as well as for the experimental wasm support?

Trying to figure out which things should be part of initial v0.1 release and which are OK to deal with later.

**Pull request closed without merging**